### PR TITLE
Regression: Fix edit check in backend articles manager, always denying edit after soft deny

### PR DESCRIPTION
**Pull request by ggppdk**
Pull Request for Issue #11463  and #11510 
#### Summary of Changes

Fixed edit check in backend not allowing edit via edit.own after soft deny
Same code used for frontend
#### Testing Instructions

(check is the same for frontend / backend)

at backend articles manager, 
at frontend views that show the edit button

make sure that people that should be allowed to edit can do and others can not

### CODE REVIEW

**Comment by infograf768** [comment]
I have tested this item :white_check_mark: successfully on <br /><br />Looks like working here.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11511">issues.joomla.org/joomla-cms/11511</a>.</sub>

**Comment by alikon** [comment]
i have experieced this issue on  GsoC 2016 Multilingual project.

test conditions

- Create a user "test" added to "Administrator" group

- Go to com_content Options, Permissions tab and disable "Edit" for "Administrator" group
permission


- Create a new content item with the "test" user 

before patch
 You can't edit your own. 

Apply patch
you can edit your own items.

**Comment by jeckodevelopment** [comment]
I have tested this item :white_check_mark: successfully on <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11511">issues.joomla.org/joomla-cms/11511</a>.</sub>

**Comment by ggeorgg** [comment]
I have tested this item :white_check_mark: successfully on 
Thanks a lot!

**Comment by zero-24** [comment]
RTC. Thanks<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11511">issues.joomla.org/joomla-cms/11511</a>.</sub>

**Comment by matrix630307** [comment]
I have tested this item successfully. Thank's.

File: administrator/components/com_content/controllers/article.php [hunk 1]
```diff
@@ -85,7 +85,8 @@ protected function allowEdit($data = array(), $key = 'id')
 		$recordId = (int) isset($data[$key]) ? $data[$key] : 0;
 		$user = JFactory::getUser();
 
-		// For new record (id:0) return component permission
+		// Zero record (id:0) return component permission, e.g. show edit btn
```
Comment (brianteeman) [inline administrator/components/com_content/controllers/article.php]: That makes less sense than it did before. Now I domt have a clue what it means

Comment (ggppdk) [inline administrator/components/com_content/controllers/article.php]: **It was like this before**,

why any code calling allowEdit() on zero id ?
only purpose would be to get component edit

for me it is meaningless, and from what i see in the code ,
**we can probably return false and break nothing**, 

I have checked this, 
- there is no code in Joomla that uses allowEdit() to decide creating new records, or to get component permissions**

so we can change it to false !

(maybe some 3rd party extends the class though and use it to get component edit, but i think it is very unlikely)

Comment (ggppdk) [inline administrator/components/com_content/controllers/article.php]: Sorry, with like this before 
- i mean current "frontend" controller
- and backend controller in J3.6.0

**Comment by zero-24** [comment]
Put off RTC so we get a retest after the last changes Thanks.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11511">issues.joomla.org/joomla-cms/11511</a>.</sub>

**Comment by ggppdk** [comment]
yes i made 1 more change (besides comments),

- always return false if article id is 0

I searched through all calls of the method, i did not find any usage that will need to return component permissions, in such a case,

so better return false (aka deny), in case someone in the future writes code to misuse this

**Comment by mbabker** [comment]
Except this is wrong.  You've introduced a hard deny rule instead of continued use of the ACL system in this scenario.  So there needs to be a pretty freakin' strong validation against this change because right now that screams massive B/C break to me.

File: administrator/components/com_content/controllers/article.php [hunk 2]
```diff
@@ -84,45 +84,32 @@ protected function allowEdit($data = array(), $key = 'id')
 	{
 		$recordId = (int) isset($data[$key]) ? $data[$key] : 0;
 		$user = JFactory::getUser();
-		$userId = $user->get('id');
 
-		// If we get a deny at the component level, we cannot override here.
-		if (!parent::allowEdit($data, $key))
+		// Zero record (id:0) return false, if caller wants component permissions just get them directly
```
Comment (mbabker) [inline administrator/components/com_content/controllers/article.php]: So a method that is supposed to check the ACL now has a hardcoded deny rule in place when it should be using the ACL system?  Am I the only one who sees an issue with this?

File: components/com_content/controllers/article.php [hunk 3]
```diff
@@ -103,45 +103,36 @@ protected function allowAdd($data = array())
...
 	{
 		$recordId = (int) isset($data[$key]) ? $data[$key] : 0;
-		$user     = JFactory::getUser();
-		$userId   = $user->get('id');
-		$asset    = 'com_content.article.' . $recordId;
+		$user = JFactory::getUser();
 
-		// Check general edit permission first.
-		if ($user->authorise('core.edit', $asset))
+		// Zero record (id:0) return false, if caller wants component permissions just get them directly
```
Comment (mbabker) [inline components/com_content/controllers/article.php]: Same issue here.  This logic is massively flawed.  We should NOT be introducing hard deny rules in ANY method which should be reading from the ACL system.  The previous behavior of reverting to the component permissions in the case of not having a record ID should be restored.

**Comment by ggppdk** [comment]
Can you find a place that is needed today ?
And something that gets broken ?

example

```
	protected function allowSave($data, $key = 'id')
	{
		$recordId = isset($data[$key]) ? $data[$key] : '0';

		if ($recordId)
		{
			return $this->allowEdit($data, $key);
		}
		else
		{
			return $this->allowAdd($data);
		}
	}
```

so calling allowEdit() on zero record id as an indirect way to get return parent controler permissions ?
which return the component permission,

is weird usage

**i will revert,** i made this change worrying of someone misusing it in the future
it creates no problems

**Comment by mbabker** [comment]
I don't have a place off hand that will break with that hardcoded change but the point of it is that something that should be reading the ACL system should be doing that, not introducing an arbitrary rule based on some condition and not even implement it consistently.

**Comment by ggppdk** [comment]
I restored the code it is same as before, when all testing was done

but that deny would break nothing anyway, because it is not used anywhere

anyway this is PR makes the allowEdit how it should have been in the first place

**Comment by mbabker** [comment]
I agree it shouldn't be getting used, but in the off case someone's code is screwed up somewhere it should still respect the ACL system or throw an Exception because at that point it's dealing with an invalid record anyway.

**Comment by ggppdk** [comment]
The code is correct to be there
i am not saying something different

but after 1 or 2 years someone may write code to misuse it
- that was my only reason for making that change

**Comment by mbabker** [comment]
It needs to be a separate proposal if you're going to push for it.  We all have a very bad habit of throwing unrelated items into pull requests because we're already making a change or want to enforce some arbitrary standard.  That crap needs to stop.

**Comment by ggppdk** [comment]
> It needs to be a separate proposal if you're going to push for it. We all have a very bad habit of throwing unrelated items into pull requests because we're already making a change or want to enforce some arbitrary standard. That crap needs to stop.

yes correct, it is unrelated to this PR, but you know how it is with finding people to test PRs etc

**Comment by mbabker** [comment]
Oh I know, I know.  I've got cache PRs stalled out because someone asked for an arbitrary extra change, one of which actually adds support for a PHP 5.3 feature to our callback system.

**Comment by infograf768** [comment]
where do we stand here?

**Comment by ggppdk** [comment]
Hello

please click on each of the commit made August 10,
https://github.com/joomla/joomla-cms/pull/11511/commits

5 commits
- 1 commit changes comments
- 2 next commits changed in the 2 controllers (frontend / backend identical), for the case of record ID zero

```
  		if (!$recordId)
 		{
-			return parent::allowEdit($data, $key);
+			return false;
 		}
```

this should be still be RTC

- **2 more commits changed it back to be the same as when all tests where made**

```
 		if (!$recordId)
 		{
-			return false;
+			return parent::allowEdit($data, $key);
 		}
```

**Comment by zero-24** [comment]
Back to RTC. Thanks<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11511">issues.joomla.org/joomla-cms/11511</a>.</sub>

**Comment by Sieger66** [comment]
I have tested this item :white_check_mark: successfully on 84d0dad8e0c5af80e598dc6db20525ba4e8d2d41<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11511">issues.joomla.org/joomla-cms/11511</a>.</sub>

**Comment by Sieger66** [comment]
I have tested this item :white_check_mark: successfully on 84d0dad8e0c5af80e598dc6db20525ba4e8d2d41<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11511">issues.joomla.org/joomla-cms/11511</a>.</sub>

**Pull request merged**